---
title: "The Fruitfly Project"
author: "Rebecca Barter"
output: html_document
---

# The *Drasophila* fruitfly project

Throughout this book we will embed many of the concepts we are examining within a real-life project, which we will call "the fruitfly project". The fruitfly project is a genomics project that I have been involved in with several of my students and my collaborators at Lawrence Berkeley National Laboratory (LBNL). <FONT COLOR="red">How did I get involved in this project? How was I introduced to Erwin and co?</FONT>. As most do, the fruitfly project began with a question: "Is spatial information useful for making gene-gene interaction discoveries?". To expand and de-code this question for the non-biologists, what the group was really asking is whether it is possible to use image data (for example images of stained drasophila embryos) to identify if . The staining process involves staining using dye certain products (typically proteins) which are highlights physically where in the embryo a particular gene is being expressed.  With this question in mind, the group of biologists at LBNL had generated 1640 spatial gene expression images <FONT COLOR="red">I need an image!</FONT> representing 755 genes for early *Drasophila* embryos. In <FONT COLOR="red">the figurer</FONT>, the gene of interest is being expressed at each location where the blue stain is visible. Each processed expression image can be represented by a numeric vector with 405 entries, one for each of the 405 pixels inside the ellipse template.



1. What the project is all about? What is our question? Bring back to "data wisdom".

1. How did you get involved in the project?

1. How was the data collected?

1. What was the process that lead to using the non-negative matrix factorization? Was that the first approach? Did you first try things that didn't work?

