<!doctype html>
<html>
<head>
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="chrome=1">
<title>Statistical Models: Theory and Applications</title>

<link rel="stylesheet" href="stylesheets/styles.css">
<link rel="stylesheet" href="stylesheets/pygment_trac.css">
<!--	<link href='http://fonts.googleapis.com/css?family=Fanwood+Text:400,400italic' rel='stylesheet' type='text/css'> -->
<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
<!--[if lt IE 9]>
 <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
 <![endif]-->
</head>
<body>
<div class="wrapper">
<header>
<h1><a href="index.html">Statistical Models: Theory and Applications</a> <h1><a><medium>Bin Yu and Rebecca Barter</medium></a></h1></h1>
<p></p>

<p class="view"><a href="https://github.com/rlbarter/Stat_Models_Book">View the Project on GitHub <small>rlbarter/Stat_Models_Book</small></a></p>




</header>










<section>  
<h1>The <em>Drosophila</em> fruitfly project</h1>

<p>Throughout this book we will embed many of the concepts we are discussing within a real-life project, which we will call &quot;the fruitfly project&quot;. The fruitfly project is a genomics project that I have been involved in with several of my students and my collaborators at Lawrence Berkeley National Laboratory (LBNL). <FONT COLOR="red">How did I get involved in this project? How was I introduced to Erwin and co?</FONT>. </p>

<p>As most do, the fruitfly project began with a question: &quot;Is spatial information useful for making gene-gene interaction discoveries?&quot;. To expand and de-code this question for the non-biologists, what the group was really asking is whether it is possible to use stained images (more specifically, images of Drosophila embryos) to identify whether genes interact with one another. For an individual embryo, the staining process involves using dye to stain specific products that are produced when a particular gene is being expressed (these products are typically proteins). The upshot of the staining process is that areas within the embryo where the stain is visible correspond to physical locations within the embryo where the gene of interest is being expressed. </p>

<p>With this question in mind, the group of biologists at LBNL had generated 1640 spatial gene expression images <FONT COLOR="red">I need an example image!</FONT> representing 755 genes for early <em>Drosophila</em> embryos. In each image, the gene of interest is being expressed at the locations within the embryo where the blue stain is visible. Each processed expression image can be represented by a numeric vector with 405 entries, one for each of the 405 pixels inside the ellipse template.</p>

<p><img src="http://insitu.fruitfly.org/insitu_image_storage/img_dir_23/insitu23270.jpe" alt="Fruitfly Embryo" style="width:304px;height:228px;"></p>

<p><FONT COLOR="red"> So the biologists had the data and the question but no idea how to answer it? They then invited Bin and co. to offer statistical insights whereby after much trial and error, we arrived at the non-negative matrix factorization approach. This approach allowed us to derive 21 principal patterns (PPs) using which we constructed spatially local correlation networks for all patterned transcroption factors during early Drosophila development</p>

<ol>
<li><p>What the project is all about? What is our question? Bring back to &quot;data wisdom&quot;.</p></li>
<li><p>How did the LBNL group get around to asking the question of &quot;can spatial information be used to make gene-gene interaction discoveries?&quot;</p></li>
<li><p>How did you get involved in the project?</p></li>
<li><p>How was the data collected?</p></li>
<li><p>What had the LBNL group tried (in terms of analysis) before Bin was brought in?</p></li>
<li><p>What was the process that lead to using the non-negative matrix factorization? Was that the first approach? Did you first try things that didn&#39;t work?</p></li>
<li><p>Could you explain clearly how we can use the spatial data to identify whether genes interact with one another?</p></li>
</ol>

</section>

<footer>

<p><small>Site powered by <a href="http://jekyllrb.com/">Jekyll, <a href="http://pandoc.org/">pandoc and <a href="http://yihui.name/knitr/">knitr</a>.</small></p>
</footer>


</div>
<script src="javascripts/scale.fix.js"></script>
<script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
})(window,document,'script','//www.google-analytics.com/analytics.js','ga');
ga('create', 'UA-39541792-2', 'github.com');
ga('send', 'pageview');
</script>    
</body>
</html>